### PR TITLE
Add SDSC marker for per-core symbolic start addresses

### PR DESCRIPTION
**Pull request by Vivek1258**
#### What type of PR is this?

- [x] feature

#### What this PR does:

Adds the SDSC-side marker for per-core symbolic start addresses (part of #2289).

When a tensor is split across cores on a symbolic dim, each core's start offset should scale with the runtime dim size (core 0 excepted, its offset is zero). Today those offsets are baked as constants from the warmup shape, so they are correct only when the runtime size equals the warmup value. For any other admissible size, cores past core 0 access the wrong HBM slice (wrong reads on split inputs, wrong writes on split outputs). Buffer size and strides stay at max (#2434); only the per-core start addresses move to actual size (#2289).

This PR adds a new `kernel_derived_symbolic` symbol kind and tags those cores with it, so a follow-up can emit the actual per-core arithmetic. At the SDSC layer the per-core entry is a negative symbol id under `isStartAddrSymbolic_`. Only the dim symbol goes into `inputSymbolsAndTags_`; `symbolDefinitions_` stays empty for now, since the per-core symbols get their definitions from the bundle arm follow-up, not the SDSC.

This is intentionally the SDSC layer only, so it can be reviewed and tested on its own.

#### Which issue(s) this PR is related to:

Part of #2289, Epic #43

#### Special notes for your reviewer:

- The follow-up bundle arm emits `base + core * ceildiv(S, split) * per_element_stride` over the dim symbol declared as a bundle `input_arg` by #2500 (PR #3238). Preferred merge order, for coherence with the bundle arm, is #3238 first, then this, then the bundle arm. This PR does not depend on #3238 and its tests pass on main, so the order is a preference, not a requirement.
- `is_derived` is kept False for the new kind so the existing `kernel_derived` bundle branch does not pick it up before the arm exists.
- Verified on a Pod: single-input (gelu) and two-input (add) symbolic runs both produce SDSCs with per-core symbol blocks, no id collisions. The detailed test scripts and the results can be found here : 
[pod_tests.zip](https://github.com/user-attachments/files/30170276/pod_tests.zip)

#### Does this PR introduce a user-facing change?

A plan that both splits a symbolic dim across cores and tiles that same dim now raises `Unsupported` at SDSC generation, instead of producing a wrong per-core address. No other user-facing change.

#### Additional note:

Tests are at the SDSC JSON level in `tests/inductor/test_codegen.py` (the marker, the split predicates, and an end-to-end `compile_op_spec` check). End-to-end execution still aborts in DeepTools until #3238 and the bundle arm land, which is expected and out of scope here.

### CODE REVIEW

**Comment by Vivek1258** [comment]
Test Results :

UT :

image

Quick repr:

```
import torch
import torch._dynamo as dynamo
import torch.nn.functional as F

DEVICE = torch.device("spyre")
torch.manual_seed(0xAFFE)


def gelu_fn(a):
    return F.gelu(a)


x = torch.rand(560, 1024, dtype=torch.float16)
x_device = x.to(DEVICE)

# Mark dim 0 as dynamic with an explicit upper bound.
# Granularity = 56, max = 616. Admissible runtime values: {56, 112, ..., 616}.
dynamo.mark_dynamic(x_device, 0, min=56, max=616)
compiled_fn = torch.compile(gelu_fn)

compiled_result = compiled_fn(x_device).cpu()
print(f"Spyre compiled result\n{compiled_result}")
```


Validate generated SDSC via 
[check_sdsc.py](https://github.com/user-attachments/files/30170311/check_sdsc.py)

Validation results :

```
(dt-inductor) [nethrakhandige@nethrakhandige-spyre-dev-pf vivek_tests]$ python3 check_sdsc.py /tmp/torchinductor_nethrakhandige/tmpjq0lymf6/inductor-spyre/sdsc_fused_gelu_0_umkvw87p/sdsc_0.json
OK   top-level has exactly one opfunc key (got ['0_gelufwd'])
     inspecting opfunc: 0_gelufwd
OK   inputSymbolsAndTags_ is non-empty dict (got {'-1': 's97'})
OK   inputSymbolsAndTags_ has NO per-core address (_addr) entries; per-core addresses are derived symbols, not inputs (got [])
OK   at least one dim symbol present (got ['s97'])
OK   exactly one dsc (got 1)
OK   dimToSymbolMapping_ non-empty (got {'mb': [-1]})
OK   all dim symbol IDs are negative (got {-1})
OK   at least one HBM allocate node (got 2)
OK   HBM allocate node 0 (allocate-Tensor0_hbm) has isStartAddrSymbolic_=1 (got 1)
OK   HBM allocate node 0 has non-empty data_ (got {'[0, 0, 0]': '-2', '[1, 0, 0]': '-3', '[2, 0, 0]': '-4', '[3, 0, 0]': '-5', '[4, 0, 0]': '-6', '[5, 0, 0]': '-7', '[6, 0, 0]': '-8', '[7, 0, 0]': '-9', '[8, 0, 0]': '-10', '[9, 0, 0]': '-11', '[10, 0, 0]': '-12', '[11, 0, 0]': '-13', '[12, 0, 0]': '-14', '[13, 0, 0]': '-15', '[14, 0, 0]': '-16', '[15, 0, 0]': '-17', '[16, 0, 0]': '-18', '[17, 0, 0]': '-19', '[18, 0, 0]': '-20', '[19, 0, 0]': '-21', '[20, 0, 0]': '-22', '[21, 0, 0]': '-23', '[22, 0, 0]': '-24', '[23, 0, 0]': '-25', '[24, 0, 0]': '-26', '[25, 0, 0]': '-27', '[26, 0, 0]': '-28', '[27, 0, 0]': '-29'})
OK   HBM allocate node 1 (allocate-Tensor1_hbm) has isStartAddrSymbolic_=1 (got 1)
OK   HBM allocate node 1 has non-empty data_ (got {'[0, 0, 0]': '-30', '[1, 0, 0]': '-31', '[2, 0, 0]': '-32', '[3, 0, 0]': '-33', '[4, 0, 0]': '-34', '[5, 0, 0]': '-35', '[6, 0, 0]': '-36', '[7, 0, 0]': '-37', '[8, 0, 0]': '-38', '[9, 0, 0]': '-39', '[10, 0, 0]': '-40', '[11, 0, 0]': '-41', '[12, 0, 0]': '-42', '[13, 0, 0]': '-43', '[14, 0, 0]': '-44', '[15, 0, 0]': '-45', '[16, 0, 0]': '-46', '[17, 0, 0]': '-47', '[18, 0, 0]': '-48', '[19, 0, 0]': '-49', '[20, 0, 0]': '-50', '[21, 0, 0]': '-51', '[22, 0, 0]': '-52', '[23, 0, 0]': '-53', '[24, 0, 0]': '-54', '[25, 0, 0]': '-55', '[26, 0, 0]': '-56', '[27, 0, 0]': '-57'})
OK   per-core symbolic addresses present (core >= 1 IDs negative)
OK   per-core address IDs do not collide with dim symbol IDs (collisions: set())
OK   per-core address symbols are NOT registered as input symbols (unexpectedly present in inputSymbolsAndTags_: set())
OK   symbolDefinitions_ is empty on this SDSC-only branch (got {})
OK   dataStageParam_['0'].ss_.symbolicDimInfo_ non-empty (got {'mb': {'maxSize_': 22, 'granularity_': 2}})
OK   dataStageParam_['0'].el_.symbolicDimInfo_ non-empty (got {'mb': {'maxSize_': 22, 'granularity_': 2}})

PASS  /tmp/torchinductor_nethrakhandige/tmpjq0lymf6/inductor-spyre/sdsc_fused_gelu_0_umkvw87p/sdsc_0.json
(dt-inductor) [nethrakhandige@nethrakhandige-spyre-dev-pf vivek_tests]$ 
```

Detailed test results and test files for single-input (gelu) and two-input (add) symbolic can be found here [pod_tests.zip](https://github.com/user-attachments/files/30170276/pod_tests.zip)

**Review by dgrove-oss — APPROVED** [review]
APPROVE — the SDSC-layer bookkeeping is correct, well-tested for the happy paths, and the registration/lookup symmetry holds. The one real gap (missing test for the Unsupported gate) is worth addressing but not blocking given the PR's narrow, explicitly-scoped nature and the fact that the gate fails safe (raises rather than silently computing a wrong address).
